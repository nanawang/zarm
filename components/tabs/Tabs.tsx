import React, { PureComponent, CSSProperties } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import TabPanel from './TabPanel';
import Carousel from '../carousel';

const getSelectIndex = (children) => {
  let selectIndex;
  React.Children.forEach(children, (item: TabPanel, index) => {
    if (item.props && item.props.selected) {
      selectIndex = index;
    }
  });
  return selectIndex;
};

export interface TabsProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Tabs extends PureComponent<TabsProps, any> {
  static Panel: typeof TabPanel;

  private carousel;

  static defaultProps = {
    prefixCls: 'za-tabs',
    disabled: false,
    canSwipe: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      value: props.value || props.defaultValue || getSelectIndex(props.children) || 0,
      prevValue: null,
    };
  }

  static getDerivedStateFromProps(nextProps, state) {
    if ('value' in nextProps && nextProps.value !== state.prevValue) {
      return {
        value: nextProps.value,
        prevValue: nextProps.value,
      };
    }
    return null;
  }

  onTabChange = (value) => {
    const { onChange } = this.props;
    if (!('value' in this.props)) {
      this.setState({ value });
    }
    typeof onChange === 'function' && onChange(value);
  };


  onTabClick = (tab, index) => {
    const { disabled, canSwipe } = this.props;
    if (disabled || tab.props.disabled) {
      return;
    }
    if (canSwipe) {
      this.carousel.onSlideTo(index);
      return;
    }
    this.onTabChange(index);
  };

  renderTabs = (tab, index) => {
    const { prefixCls, disabled } = this.props;
    const itemCls = classnames(`${prefixCls}__tab`, tab.props.className, {
      [`${prefixCls}__tab--disabled`]: disabled || tab.props.disabled,
    });

    return (
      <li
        role="tab"
        key={+index}
        className={itemCls}
        onClick={() => this.onTabClick(tab, index)}
      >
        {tab.props.title}
      </li>
    );
  };

  render() {
    const { prefixCls, className, lineWidth, canSwipe, children, disabled } = this.props;
    const { value } = this.state;
    const classes = classnames(prefixCls, className);

    // 渲染选项
    const tabsRender = React.Children.map(children, this.renderTabs);

    // 渲染内容
    let contentRender;

    if (canSwipe) {
      contentRender = (
        <Carousel
          swipeable={!disabled}
          direction="left"
          showPagination={false}
          activeIndex={value}
          ref={(ele) => { this.carousel = ele; }}
          onChange={(v) => this.onTabChange(v)}
        >
          {React.Children.map(children, (item: TabPanel) => <div>{item.props.children}</div>)}
        </Carousel>
      );
    } else {
      contentRender = React.Children.map(children, (item: TabPanel, index) => {
        return item.props.children && <TabPanel {...item.props} selected={value === index} />;
      });
    }

    const lineStyle: CSSProperties = {
      width: `${100 / React.Children.count(children)}%`,
      left: `${(value / React.Children.count(children)) * 100}%`,
      // right: `${(React.Children.count(children) - this.state.value - 1) / React.Children.count(children) * 100}%`,
      // transition: `right 0.3s cubic-bezier(0.35, 0, 0.25, 1), left 0.3s cubic-bezier(0.35, 0, 0.25, 1) 0.09s`,
    };

    let lineInnerRender;
    if (lineWidth) {
      lineStyle.backgroundColor = 'transparent';
      lineInnerRender = <span className={`${prefixCls}__line__inner`} style={{ width: lineWidth }} />;
    }

    return (
      <div className={classes}>
        <div className={`${prefixCls}__header`}>
          <ul role="tablist">{tabsRender}</ul>
          <div className={`${prefixCls}__line`} style={lineStyle}>{lineInnerRender}</div>
        </div>
        <div className={`${prefixCls}__body`}>
          {contentRender}
        </div>
      </div>
    );
  }
}
