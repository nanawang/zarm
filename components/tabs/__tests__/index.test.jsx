import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tabs from '../index';

describe('Tab', () => {
  it('renders correctly', () => {
    const wrapper = render(
      <Tabs>
        <Tabs.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tabs.Panel>
        <Tabs.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tabs.Panel>
      </Tabs>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('canSwipe', () => {
    const wrapper = render(
      <Tabs canSwipe>
        <Tabs.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tabs.Panel>
        <Tabs.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tabs.Panel>
      </Tabs>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('lineWidth is auto', () => {
    const wrapper = shallow(
      <Tabs lineWidth="auto">
        <Tabs.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tabs.Panel>
        <Tabs.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tabs.Panel>
      </Tabs>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ lineWidth: 30 });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('receive new value', () => {
    const onChange = jest.fn();
    const wrapper = shallow(
      <Tabs onChange={onChange}>
        <Tabs.Panel title="选项卡1">
          <div>试试点我左滑</div>
        </Tabs.Panel>
        <Tabs.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tabs.Panel>
      </Tabs>,
    );
    wrapper.setProps({ value: 1 });
  });

  it('click tabs', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <Tabs canSwipe onChange={onChange}>
        <Tabs.Panel title="选项卡1" disabled>
          <div>试试点我左滑</div>
        </Tabs.Panel>
        <Tabs.Panel title="选项卡2">
          <div>试试点我右滑</div>
        </Tabs.Panel>
      </Tabs>,
    );
    wrapper.find('.za-tabs__tab').first().simulate('click');
    wrapper.find('.za-tabs__tab').last().simulate('click');
    expect(onChange).toBeCalledWith(1);
  });
});
