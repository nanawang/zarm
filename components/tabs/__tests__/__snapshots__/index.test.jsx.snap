// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab canSwipe 1`] = `
<div
  class="za-tabs"
>
  <div
    class="za-tabs__header"
  >
    <ul
      role="tablist"
    >
      <li
        class="za-tabs__tab"
        role="tab"
      >
        选项卡1
      </li>
      <li
        class="za-tabs__tab"
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      class="za-tabs__line"
      style="width:50%;left:0%"
    />
  </div>
  <div
    class="za-tabs__body"
  >
    <div
      class="za-carousel za-carousel--horizontal"
    >
      <div
        class="za-carousel__items"
      >
        <div
          class="za-carousel__item"
        >
          <div>
            试试点我左滑
          </div>
        </div>
        <div
          class="za-carousel__item"
        >
          <div>
            试试点我右滑
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tab lineWidth is auto 1`] = `
<div
  className="za-tabs"
>
  <div
    className="za-tabs__header"
  >
    <ul
      role="tablist"
    >
      <li
        className="za-tabs__tab"
        key="0/.0"
        onClick={[Function]}
        role="tab"
      >
        选项卡1
      </li>
      <li
        className="za-tabs__tab"
        key="1/.1"
        onClick={[Function]}
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      className="za-tabs__line"
      style={
        Object {
          "backgroundColor": "transparent",
          "left": "0%",
          "width": "50%",
        }
      }
    >
      <span
        className="za-tabs__line__inner"
        style={
          Object {
            "width": "auto",
          }
        }
      />
    </div>
  </div>
  <div
    className="za-tabs__body"
  >
    <TabPanel
      key=".0"
      prefixCls="za-tabs"
      selected={true}
      title="选项卡1"
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key=".1"
      prefixCls="za-tabs"
      selected={false}
      title="选项卡2"
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
  </div>
</div>
`;

exports[`Tab lineWidth is auto 2`] = `
<div
  className="za-tabs"
>
  <div
    className="za-tabs__header"
  >
    <ul
      role="tablist"
    >
      <li
        className="za-tabs__tab"
        key="0/.0"
        onClick={[Function]}
        role="tab"
      >
        选项卡1
      </li>
      <li
        className="za-tabs__tab"
        key="1/.1"
        onClick={[Function]}
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      className="za-tabs__line"
      style={
        Object {
          "backgroundColor": "transparent",
          "left": "0%",
          "width": "50%",
        }
      }
    >
      <span
        className="za-tabs__line__inner"
        style={
          Object {
            "width": 30,
          }
        }
      />
    </div>
  </div>
  <div
    className="za-tabs__body"
  >
    <TabPanel
      key=".0"
      prefixCls="za-tabs"
      selected={true}
      title="选项卡1"
    >
      <div>
        试试点我左滑
      </div>
    </TabPanel>
    <TabPanel
      key=".1"
      prefixCls="za-tabs"
      selected={false}
      title="选项卡2"
    >
      <div>
        试试点我右滑
      </div>
    </TabPanel>
  </div>
</div>
`;

exports[`Tab renders correctly 1`] = `
<div
  class="za-tabs"
>
  <div
    class="za-tabs__header"
  >
    <ul
      role="tablist"
    >
      <li
        class="za-tabs__tab"
        role="tab"
      >
        选项卡1
      </li>
      <li
        class="za-tabs__tab"
        role="tab"
      >
        选项卡2
      </li>
    </ul>
    <div
      class="za-tabs__line"
      style="width:50%;left:0%"
    />
  </div>
  <div
    class="za-tabs__body"
  >
    <div
      class="za-tabs__panel za-tabs__panel--active"
      role="tabpanel"
    >
      <div>
        试试点我左滑
      </div>
    </div>
    <div
      class="za-tabs__panel"
      role="tabpanel"
    >
      <div>
        试试点我右滑
      </div>
    </div>
  </div>
</div>
`;
